File: sdk-typescript/packages/creator/src/creator.ts
/**
 * WAVE SDK - Creator API
 *
 * Creator monetization, subscriptions, tips, and payouts.
 */

import type { WaveClient, PaginationParams, PaginatedResponse, Timestamps } from '@wave-av/core';

export interface CreatorProfile extends Timestamps {
  id: string;
  user_id: string;
  organization_id: string;
  display_name: string;
  bio?: string;
  avatar_url?: string;
  banner_url?: string;
  subscriber_count: number;
  follower_count: number;
  total_revenue_cents: number;
  verified: boolean;
  tier: "starter" | "pro" | "partner";
}
export interface Subscription {
  id: string;
  creator_id: string;
  subscriber_id: string;
  tier: string;
  price_cents: number;
  status: "active" | "cancelled" | "past_due";
  current_period_start: string;
  current_period_end: string;
  created_at: string;
}
export interface Tip {
  id: string;
  creator_id: string;
  tipper_id: string;
  amount_cents: number;
  message?: string;
  stream_id?: string;
  created_at: string;
}
export interface Payout {
  id: string;
  creator_id: string;
  amount_cents: number;
  status: "pending" | "processing" | "completed" | "failed";
  method: "bank_transfer" | "paypal" | "stripe";
  requested_at: string;
  completed_at?: string;
}
export interface RevenueReport {
  creator_id: string;
  period: string;
  total_cents: number;
  subscription_cents: number;
  tip_cents: number;
  ad_cents: number;
  platform_fee_cents: number;
  net_cents: number;
}
export interface UpdateProfileRequest {
  display_name?: string;
  bio?: string;
  avatar_url?: string;
  banner_url?: string;
}
export interface ListSubscriptionsParams extends PaginationParams {
  status?: string;
  tier?: string;
}

/**
 * Creator monetization: profiles, subscriptions, tips, and payouts.
 *
 * @example
 * ```typescript
 * const profile = await wave.creator.getProfile(creatorId);
 * const revenue = await wave.creator.getRevenue(creatorId, { period: "month" });
 * await wave.creator.requestPayout(creatorId, { amount_cents: 10000, method: "stripe" });
 * ```
 */
export class CreatorAPI {
  private readonly client: WaveClient;
  private readonly basePath = "/v1/creators";
  constructor(client: WaveClient) {
    this.client = client;
  }

  async getProfile(creatorId: string): Promise<CreatorProfile> {
    return this.client.get<CreatorProfile>(`${this.basePath}/${creatorId}`);
  }
  async updateProfile(creatorId: string, request: UpdateProfileRequest): Promise<CreatorProfile> {
    return this.client.patch<CreatorProfile>(`${this.basePath}/${creatorId}`, request);
  }
  async getRevenue(
    creatorId: string,
    params?: { period?: string; start_date?: string; end_date?: string },
  ): Promise<RevenueReport> {
    return this.client.get<RevenueReport>(`${this.basePath}/${creatorId}/revenue`, {
      params: params as Record<string, string | number | boolean | undefined>,
    });
  }
  async listSubscriptions(
    creatorId: string,
    params?: ListSubscriptionsParams,
  ): Promise<PaginatedResponse<Subscription>> {
    return this.client.get<PaginatedResponse<Subscription>>(
      `${this.basePath}/${creatorId}/subscriptions`,
      { params: params as Record<string, string | number | boolean | undefined> },
    );
  }
  async listTips(creatorId: string, params?: PaginationParams): Promise<PaginatedResponse<Tip>> {
    return this.client.get<PaginatedResponse<Tip>>(`${this.basePath}/${creatorId}/tips`, {
      params: params as Record<string, string | number | boolean | undefined>,
    });
  }
  async createTipJar(
    creatorId: string,
    config: { enabled: boolean; min_amount_cents?: number; suggested_amounts?: number[] },
  ): Promise<{ enabled: boolean }> {
    return this.client.post<{ enabled: boolean }>(`${this.basePath}/${creatorId}/tip-jar`, config);
  }
  async listPayouts(
    creatorId: string,
    params?: PaginationParams,
  ): Promise<PaginatedResponse<Payout>> {
    return this.client.get<PaginatedResponse<Payout>>(`${this.basePath}/${creatorId}/payouts`, {
      params: params as Record<string, string | number | boolean | undefined>,
    });
  }
  async requestPayout(
    creatorId: string,
    request: { amount_cents: number; method: Payout["method"] },
  ): Promise<Payout> {
    return this.client.post<Payout>(`${this.basePath}/${creatorId}/payouts`, request);
  }
  async getAnalytics(
    creatorId: string,
    params?: { period?: string },
  ): Promise<Record<string, unknown>> {
    return this.client.get<Record<string, unknown>>(`${this.basePath}/${creatorId}/analytics`, {
      params: params as Record<string, string | number | boolean | undefined>,
    });
  }
}

export function createCreatorAPI(client: WaveClient): CreatorAPI {
  return new CreatorAPI(client);
}

File: sdk-python/wave_av_sdk/creator.py
"""WAVE SDK - Creator API. Monetization, subscriptions, tips, and payouts."""
from __future__ import annotations
from typing import Any
from pydantic import BaseModel
from wave.client import WaveClient

class CreatorProfile(BaseModel):
    id: str; user_id: str; display_name: str; subscriber_count: int = 0; total_revenue_cents: int = 0; verified: bool = False; tier: str = "starter"; created_at: str; updated_at: str

class CreatorAPI:
    def __init__(self, client: WaveClient): self._client = client; self._base = "/v1/creators"
    def get_profile(self, creator_id: str) -> CreatorProfile: return CreatorProfile(**self._client.get(f"{self._base}/{creator_id}"))
    def update_profile(self, creator_id: str, **kwargs: Any) -> CreatorProfile: return CreatorProfile(**self._client.patch(f"{self._base}/{creator_id}", json=kwargs))
    def get_revenue(self, creator_id: str, **params: Any) -> dict: return self._client.get(f"{self._base}/{creator_id}/revenue", params={k: v for k, v in params.items() if v is not None})
    def list_subscriptions(self, creator_id: str, **params: Any) -> dict: return self._client.get(f"{self._base}/{creator_id}/subscriptions", params={k: v for k, v in params.items() if v is not None})
    def list_tips(self, creator_id: str, **params: Any) -> dict: return self._client.get(f"{self._base}/{creator_id}/tips", params={k: v for k, v in params.items() if v is not None})
    def create_tip_jar(self, creator_id: str, enabled: bool = True, **kwargs: Any) -> dict: return self._client.post(f"{self._base}/{creator_id}/tip-jar", json={"enabled": enabled, **kwargs})
    def list_payouts(self, creator_id: str, **params: Any) -> dict: return self._client.get(f"{self._base}/{creator_id}/payouts", params={k: v for k, v in params.items() if v is not None})
    def request_payout(self, creator_id: str, amount_cents: int, method: str) -> dict: return self._client.post(f"{self._base}/{creator_id}/payouts", json={"amount_cents": amount_cents, "method": method})
    def get_analytics(self, creator_id: str, **params: Any) -> dict: return self._client.get(f"{self._base}/{creator_id}/analytics", params={k: v for k, v in params.items() if v is not None})
